appmgr/src/update/mod.rs

```diff
@@ -0,0 +1,289 @@
+use anyhow::{anyhow, bail, Result};
+use clap::ArgMatches;
+use digest::Digest;
+use futures::Stream;
+use lazy_static::lazy_static;
+use regex::Regex;
+use rpc_toolkit::command;
+use sha2::Sha256;
+use tokio::io::AsyncWriteExt;
+use tokio::pin;
+use tokio_stream::StreamExt;
+
+use crate::context::RpcContext;
+use crate::update::latest_information::LatestInformation;
+use crate::{Error, ErrorKind, ResultExt};
+
+/// An user/ daemon would call this to update the system to the latest version and do the updates available,
+/// and this will return something if there is an update, and in that case there will need to be a restart.
+#[command(display(display_properties))]
+pub async fn update_system(#[context] ctx: RpcContext) -> Result<UpdateSystem, Error> {
+    if let None = maybe_do_update(ctx).await? {
+        return Ok(UpdateSystem::Updated);
+    }
+    Ok(UpdateSystem::NoUpdates)
+}
+
+/// What is the status of the updates?
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug)]
+pub enum UpdateSystem {
+    NoUpdates,
+    Updated,
+}
+
+fn display_properties(status: UpdateSystem, _: &ArgMatches<'_>) {
+    match status {
+        UpdateSystem::NoUpdates => {
+            println!("Updates are ready, please reboot");
+        }
+        UpdateSystem::Updated => {
+            println!("No updates needed");
+        }
+    }
+}
+
+const URL: &str = "https://beta-registry-0-3.start9labs.com/eos/latest";
+const HEADER_KEY: &str = "CHECKSUM";
+mod latest_information;
+
+enum WritableDrives {
+    Green,
+    Blue,
+}
+
+struct Boot;
+
+/// We are going to be creating some folders and mounting so
+/// we need to know the labels for those types. These labels
+/// are the labels that are shipping with the embassy, blue/ green
+/// are where the os sits and will do a swap during update.
+trait FileType {
+    fn mount_folder(&self) -> String {
+        format!("/media/{}", self.label())
+    }
+    fn label(&self) -> String;
+}
+impl FileType for WritableDrives {
+    fn label(&self) -> String {
+        match self {
+            WritableDrives::Green => "green",
+            WritableDrives::Blue => "blue",
+        }
+        .to_string()
+    }
+}
+impl FileType for Boot {
+    fn label(&self) -> String {
+        "system-boot".to_string()
+    }
+}
+
+/// Proven data that this is mounted, should be consumed in an unmount
+struct MountedResource<X: FileType>(X);
+impl<X: FileType> MountedResource<X> {
+    async fn unmount_label(&self) -> Result<()> {
```

Comment:
Would have done a drop, but async drop doesn't work well in my experience, would like a better pattern?